Shared Tree: Use minVersionForCollab instead of oldestCompatibleClient to specify min client version

TommyBrosman opened the pull request:
## Description

This change replaces `SharedTreeOptions.oldestCompatibleClient` with `minVersionForCollab`. Additionally, Shared Tree will now use the `minVersionForCollab` specified at the ContainerRuntime level when `minVersionForCollab` is not specified at the Shared Tree level.

Changes APIs and codec code to use MinimumVersionForCollab in place of FluidClientVersion. Also changes FluidClientVersion to a const object that contains MinimumVersionForCollab values.

## Breaking Changes

This change breaks the following alpha APIs:

- `SharedTreeOptions`: `oldestCompatibleClient` has been replaced with `minVersionForCollab`. See changeset notes on how to specify the minimum compatible version.
- `CodecWriteOptions.oldestCompatibleClient` (type change)
- `makeDetachedFieldIndexCodec` (parameter type change)
- `TreeAlpha.exportCompressed` (parameter type change)
- `extractPersistedSchema` (parameter type change)

## Reviewer Guidance

## Todo

- [x] Rebase onto main once #25130 is committed.
- [x] Determine which flags/formats can be set as part of this change.

TommyBrosman committed:
- Changed all usages of FluidClientVersion to MinimumVersionForCollab.
- Converted FluidClientVersion to a const object with MinimumVersionForCollab values.
- Blocked out `selectFeatureFlags` in treeFactory.ts.

CraigMacomber commented on packages/dds/tree/src/codec/codec.ts:
```diff
@@ -457,19 +452,21 @@ export enum FluidClientVersion {
 	/**
 	 * Enable unreleased and unfinished features.
 	 * @remarks
+	 * pkgVersion will always be the latest version, therefore the version that enables all features.
```
This comment could be in privateRemarks, but it should not be un the user facing documentation since the fact that it uses pkgVersion is an implementation detail.

Also note that using pkgVersion here doesn't actually work as documented, since that would only opt into things that are planned to stabilize in the next release.

I think for not it might be best to simply remove the EnableUnstableFeatures option as there isn't a good way to support it with MinimumVersionForCollab

CraigMacomber commented on packages/dds/tree/src/codec/codec.ts:
```diff
@@ -457,19 +452,21 @@ export enum FluidClientVersion {
 	// New formats introduced in 2.52:
 	// - DetachedFieldIndex FormatV2
-	v2_52 = 2.052,
+	v2_52: "2.52.0" as MinimumVersionForCollab,
```
`as` casts are relatively unsafe.

I see no reasons to type erase the exact string value here by unsafely converting to MinimumVersionForCollab. Everything should work fine if you simply remove the `as MinimumVersionForCollab` and more specific typing will be preserved that way.

If you do with to have the compiler validate the values here are compatible with MinimumVersionForCollab, you can do so in any of several safer ways, like a `satisfies Record<string, MinimumVersionForCollab>` clause.

CraigMacomber commented on packages/dds/tree/src/test/codec/versioned/codec.spec.ts:
```diff
@@ -42,16 +39,17 @@ describe("versioned Codecs", () => {
 		it("round trip", () => {
 			const codec1 = builder.build({
-				oldestCompatibleClient: 2 as FluidClientVersion,
+				oldestCompatibleClient: "2.0.0" as MinimumVersionForCollab,
```
no cast should be required or performed here.

TommyBrosman replied:
That is true for this line, though "6.0.0" needs to be casted since the regex for MinimumVersionForCollab only accepts a major version of 1 or 2.

CraigMacomber commented on packages/dds/tree/src/codec/codec.ts:
```diff
@@ -457,19 +452,21 @@ export enum FluidClientVersion {
-	EnableUnstableFeatures = Number.POSITIVE_INFINITY,
-}
+	EnableUnstableFeatures: pkgVersion as MinimumVersionForCollab,
+};
```
you don't want this object being mutable, and you should type check it:


```suggestion
} as const satisfies Record<string, MinimumVersionForCollab>;
```

TommyBrosman committed:
- Tweaked changeset to comply with Vale.
- Removed FluidClientVersion.EnableUnstableFeatures since we cannot robustly support it with MinVersionForCollab. The fuzz test code that used this value now passes pkgVersion directly.
- Made FluidClientVersion actually immutable and correctly typed.

TommyBrosman committed:
Update .changeset/warm-windows-cross.md

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>

TommyBrosman committed:
New policy for oldestCompatibleClient:
- minVersionForCollab is always used as the oldestCompatibleClient version with the exception of tests.
- SharedTreeOptions no longer allows setting oldestCompatibleClient.

TommyBrosman committed:
Update .changeset/warm-windows-cross.md

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

TommyBrosman committed:
Apply suggestion from @CraigMacomber

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

TommyBrosman committed:
Apply suggestion from @CraigMacomber

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

TommyBrosman committed:
PR feedback.

TommyBrosman committed:
- Added minVersionForCollab on SharedTreeOptions as well as CodecWriteOptions.
- Renamed all usages of oldestClientVersion to minVersionForCollab for consistency across APIs.

TommyBrosman committed:
Minor: tweaked md line length.

alexvy86 reviewed (approved):
Approving for docs. I had a quick look at the rest of the change and made sense to me but probably get someone else to also review.

TommyBrosman committed:
Update packages/dds/tree/src/treeFactory.ts

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

TommyBrosman committed:
Merge branch 'main' into set-st-flags

TommyBrosman committed:
Merge branch 'set-st-flags' of https://github.com/TommyBrosman/FluidFramework into set-st-flags

TommyBrosman committed:
Vale.

TommyBrosman committed:
Fixed tree-cli-app.

TommyBrosman committed:
Regenerated fluid-framework.alpha.api.md.
